packages/runtime/container-runtime/src/summary/runningSummarizer.ts

```diff
@@ -664,6 +661,139 @@ export class RunningSummarizer implements IDisposable {
+			const shouldRetry = (
+				failureData: ISubmitSummaryFailureResult | undefined,
+			): ShouldRetryResult => {
+				// For lastSummary, do not retry.
+				// If the failure data does not specify `retryAfterSeconds` or the stage at which summarization failed,
+				// do not retry.
+				// If maxRetries have been exceeded, do not retry anymore.
+				if (
+					reason === "lastSummary" ||
+					failureData === undefined ||
+					failureData.retryAfterSeconds === undefined ||
+					failureData.stage === undefined ||
+					currentRetries >= maxRetries
+				) {
+					return { retry: false };
+				}
+
+				// If summarization failed at the same stage as before, check if the max number of retries for this
+				// stage have been reached and if so, do not retry anymore.
+				if (failureData.stage === previousStage) {
+					if (currentRetriesForStage >= maxRetriesForStage) {
+						return { retry: false };
+					}
+					return {
+						retry: true,
+						stage: failureData.stage,
+						delayMs: failureData.retryAfterSeconds,
+					};
+				}
+
+				// If summarization failed at a different stage than before, reset the retry counters for this stage.
+				// Corner case - The first failure will be a different stage from previous stage (undefined) and in that
+				// case one try has already happened.
+				currentRetriesForStage = currentRetries === 0 ? 0 : -1;
+				// If the failure data specifies retryCount, use it. Else, use the default value.
+				maxRetriesForStage = failureData.retryCount ?? defaultMaxRetriesForStage;
+				return {
+					retry: true,
+					stage: failureData.stage,
+					delayMs: failureData.retryAfterSeconds,
+				};
+			};
+
+			let done = false;
+			do {
+				if (this.cancellationToken.cancelled) {
+					return;
+				}
+
+				currentRetries++;
+				currentRetriesForStage++;
+
+				const summarizeProps: ISummarizeTelemetryProperties = {
+					reason,
+					summaryAttempts: currentRetries,
+					...summarizeOptions,
+				};
+				const summarizeResult = this.generator.summarize(
+					summarizeProps,
+					summarizeOptions,
+					cancellationToken,
+				);
+
+				const ackNackResult = await summarizeResult.receivedSummaryAckOrNack;
+				if (ackNackResult.success) {
+					return;
+				}
+
+				const submitResult = await summarizeResult.summarySubmitted;
+				assert(
+					!submitResult.success,
+					"Summary submission should have failed for failed summarization",
+				);
+				const submitFailureResult = ackNackResult.data ?? submitResult.data;
+				const shouldRetryResult = shouldRetry(submitFailureResult);
+
+				if (!shouldRetryResult.retry) {
+					done = true;
+				} else {
+					this.mc.logger.sendPerformanceEvent({
+						eventName: "SummarizeAttemptDelay",
+						duration: shouldRetryResult.delayMs,
+						// summaryNackDelay: submitResult,
+						stage: shouldRetryResult.stage,
+						...summarizeProps,
+					});
+					await delay(shouldRetryResult.delayMs * 1000);
```

Comment:
Nit: Either delayMs should be renamed to delaySeconds or this `* 1000` should be removed

Reply:
Oh yeah, totally missed that it is already in seconds.

Reply:
Updated to `delaySeconds`